dev-tools/mage/packageversion.go

```diff
@@ -0,0 +1,109 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package mage
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+)
+
+const PackageVersionFilename = ".package-version"
+
+type packageVersion struct {
+	Version      string `json:"version"`
+	BuildID      string `json:"build_id"`
+	ManifestURL  string `json:"manifest_url"`
+	SummaryURL   string `json:"summary_url"`
+	CoreVersion  string `json:"core_version"`
+	StackBuildID string `json:"stack_build_id"`
+}
+
+func initPackageVersion() error {
+	if os.Getenv("USE_PACKAGE_VERSION") != "true" {
+		return nil
+	}
+
+	_, err := os.Stat(PackageVersionFilename)
+	if err != nil {
+		if os.IsNotExist(err) {
+			log.Printf("USE_PACKAGE_VERSION is set, but %q does not exist, not overriding\n", PackageVersionFilename)
+			return nil
+		}
+		return fmt.Errorf("failed to stat %q: %w", PackageVersionFilename, err)
+	}
+
+	pv, err := readPackageVersion()
+	if err != nil {
+		// err is wrapped in readPackageVersion
+		return err
+	}
+
+	PackagingFromManifest = true
+	ManifestURL = pv.ManifestURL
+	agentPackageVersion = pv.CoreVersion
+	Snapshot = true
+
+	_ = os.Setenv("BEAT_VERSION", pv.CoreVersion)
+	_ = os.Setenv("AGENT_VERSION", pv.Version)
+	_ = os.Setenv("AGENT_STACK_VERSION", pv.Version)
+	_ = os.Setenv("SNAPSHOT", "true")
+	return nil
+}
+
+func UpdatePackageVersion(version string, buildID string, manifestURL string, summaryURL string) error {
+	packageVersion := packageVersion{
+		Version:      version,
+		BuildID:      buildID,
+		ManifestURL:  manifestURL,
+		SummaryURL:   summaryURL,
+		CoreVersion:  strings.ReplaceAll(version, "-SNAPSHOT", ""),
+		StackBuildID: fmt.Sprintf("%s-SNAPSHOT", buildID),
+	}
+
+	if err := writePackageVersion(packageVersion); err != nil {
+		// err is wrapped in writePackageVersion
+		return err
+	}
+	return nil
+}
+
+func writePackageVersion(pv packageVersion) error {
+	f, err := os.OpenFile(PackageVersionFilename, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0o644)
+	if err != nil {
+		return fmt.Errorf("failed to open %q for write: %w", PackageVersionFilename, err)
+	}
+	defer f.Close()
+
+	pvBytes, err := json.MarshalIndent(pv, "", "  ")
+	if err != nil {
+		return fmt.Errorf("failed to marshal package version: %w", err)
+	}
+
+	_, err = f.Write(pvBytes)
+	if err != nil {
+		return fmt.Errorf("failed to write package version: %w", err)
+	}
```

Comment:
We could reduce the number of lines here with `os.WriteFile`.

Reply:
take a look here d6a460c0ddd9a15a3fd2ed3ce543fa036ef322d3 🙂